accepted/2021/install-location-per-architecture.md

```diff
@@ -0,0 +1,229 @@
+If running apphost works in all configurations it will also make
+`dotnet run` work in almost all configurations, since that invokes apphost
+if the app is configured to build one.
+It can also affect `dotnet test`. Currently we only use apphost to run tests
+on Windows, to solve the x86/x64 differences. We could extend this to other platforms
+and architectures as well. In that case, `dotnet test` would not need to have
+any specific knowledge of where to find the install location, it would just
+invoke the right apphost.
+
+On all platforms the apphost will have a default location hardcoded
+and so if the installer puts the layout into that location it doesn't have to
+register that installation. But it would be cleaner if it did so anyway.
+Currently we always register the installation on Windows, but we don't seem
+to do so on Linux and macOS.
+
+*Note: If installers write the file, they should also read it and validate
+that they don't overwrite it with different information. It's unclear yet
+if this is necessary and in which situations.*
+
+#### Support for downlevel apphost
+
+In order to support running downlevel apphost apps on multi-arch non-Windows
+architectures (currently only applies to macOS), the installer of .NET 6 and above
+should write the configuration file in a specific shape, for example (on macOS arm64):
+
+```console
+/usr/local/share/dotnet/x64
+arm64=/usr/local/share/dotnet
+x64=/usr/local/share/dotnet/x64
+```
+
+With the above proposed changes this would guarantee that all apphosts
+would work if the right frameworks are installed.
+
+* 3.1 and 5 apphost would only read the first line, which points to x64 install.
+Since 3.1 and 5 are only supported on x64 on macOS, that is the right location.
+* arm64 6 apphost would react to the `arm64` location
+* x64 6 apphost would react to the `x64` location
+
+### Evolve environment variable overrides
+
+Similarly to the registration mechanism, the environment variable mechanism
+doesn't support multi-arch on Linux and macOS at all. And the existing support
+on Windows only works for x86 and x64.
+
+#### Proposed architecture specific environment variables
+
+Apphost should first recognize architecture specific environment variable
+in the format `DOTNET_ROOT_ARCH`. So for example `DOTNET_ROOT_x64` or `DOTNET_ROOT_ARM64`.
+The same mechanism would be used on all OSes, this means that x86 apphost
+on Windows would recognize `DOTNET_ROOT_X86` (it would also recognize
+`DOTNET_ROOT(x86)` as it does today for backward compat reasons).
+
+If the architecture specific environment variable is not set all apphosts
+should fallback to the default `DOTNET_ROOT` environment variable.
+
+By far the most common use case for environment variable overrides
+is usage of private install locations. In that case it's almost always
+the case that only one architecture is needed. For these cases the existing
+`DOTNET_ROOT` can be used. Going forward we may recommend to switch
```

Comment:
I do not think we would want to recomend people using the arch specific env variables for private installations unless they really have the scenario that needs them. It would unnecessarily complicate every other build script that uses the private install.

Reply:
Yeah, I think `DOTNET_ROOT` would be preferred. Arch-specific is probably only necessary for indirect invocations that you don't control, and need to point to a local dotnet copy.

Reply:
Thanks - I updated the wording accordingly